start: clarify existing content (#265)
iterative/mlem.ai: content/docs/get-started/index.md

---

> description: 'Learn how you can use MLEM to easily manage and deploy machine learning models'

Author: Spelling out "machine learning" and using "ML" here and there for better SEO.

---
description:
  'Learn how to use MLEM to manage and deploy machine learning models easily.'
---

# Get Started with MLEM

> <!--
> ## Get Started with MLEM
> -->

Reviewer: ```suggestion
```

Author: This is a trick so that the header is shown in the ToC (see pic below). I've been using these in DVC docs as a workaround to https://github.com/iterative/gatsby-theme-iterative/issues/37 for now -- we can batch remove them when/if that gets addressed.

<img width="822" alt="image" src="https://user-images.githubusercontent.com/1477535/209417153-a5773641-e2c2-4dda-b5d1-4685f6137b36.png">

> <admon type="info">

Reviewer: This looks strange right at the top of the tutorial. Why do we need this admon?

Author: Well, it's 3rd-party tool documentation, so it should be separated somehow IMO. Would you prefer a hidden `<details>` section? I wouldn't hide it since the steps are required to follow the tutorial. Currently it looks like this:

<img width="1234" alt="image" src="https://user-images.githubusercontent.com/1477535/209610960-6ff551b1-e5dc-4cd6-911d-928cbe22aa56.png">

Author: > p.s. maybe if the admonition seems to loud to you we can use an old-school block quote (like this comment).

)
```

Here, we loaded a well-known [Iris flower dataset] with scikit-learn and trained
a simple classifier (see the full list of [supported ML frameworks]). Instead of

> pickling the model, we captured it using `mlem.api.save`, which created the

Author: "To persist" doesn't really mean [what we want it to](https://github.com/iterative/dvc.org/issues/2960). "Save" was better anyway (it's simple and it matches the MLEM fn name).

)
```

> Here, we loaded a well-known [Iris flower dataset] with scikit-learn and trained

Reviewer: ```suggestion
Here, we loaded a well-known [Iris flower dataset] with scikit-learn and train
```

Author: This seems correct atm. We're using simple past tense in most places after code blocks, same as here ("we loaded ... and trained ...") 🙂  Resolivng for now (please lmk if I didn't understand the suggestion).

a simple classifier (see the full list of [supported ML frameworks]). We then

> captures it using `mlem.api.save`, which creates the following files:

Reviewer: ```suggestion
capture it using `mlem.api.save`, which creates the following files:
```

Reviewer: Why not "persist"? "capture" sounds strange to me (may be just my bad English 😅 )

Author: "to persist" doesn't mean what we want it to 🙂 

See discussion in https://github.com/iterative/dvc.org/issues/2960

```
models
├── rf
└── rf.mlem
```

Alongside the model binary `models/rf`, MLEM saved metafile `models/rf.mlem`,
which contains all the information needed to use the model later:

1. Model methods: Like `predict` and `predict_proba`
2. Input data schema: Describes the dataframe (Iris dataset)
3. Python Requirements: `sklearn` and `pandas` in this case, with the specific
   versions used to train the model

> <admon type="info">

Reviewer: 💡 was looking more appealing to me than ℹ️ . Just complaining, if that's our standard for all docs websites, please ignore

Author: Well, is this a tip or is it useful information? The way each one looks could be changed, but that's a design question IMO. No strong opinion though, feel free to revert this if you prefer.

</details>

<details>

### Click to learn about model codification

#### Why do it the MLEM way ?

Saving models to files or loading them back into python objects may seem like a

> deceptively simple task at first. For example, `pickle` and `torch` libraries

Author: Moved all this to the User Guide (index page for now).

</details>

<details>
## Model prediction

### Click to learn about model codification

#### Why do it the MLEM way ?

Saving models to files or loading them back into python objects may seem like a
deceptively simple task at first. For example, `pickle` and `torch` libraries
can serialize/deserialize model objects to/from files. However, MLEM adds some
"special sauce" by inspecting the objects and serializing their metadata into
`.mlem` files and intelligently using this later on. This metadata is necessary
to reliably enable actions like packaging and serving of different models types
down in various ways. MLEM allows us to automate a lot of the pain points we
would hit later on in our ML workflow by codifying and managing this metadata
about our models (or other objects) for us.

> </details>

Reviewer: My take: I'd keep this section. It's a bit too abstract for my taste, so I can suggest putting an example here. Saying something like:
- we need to know there is predict and predict_proba to build a REST API interface. Same for input data schema.
- we need to know it's pandas and numpy to pip install requirements in containers on Heroku/AWS

I'd keep it - with examples it can be easier to understand why MLEM is powerful.
WDYT?

Author: > I'd keep this section 

If we want to present the GS along these lines (give more motivation) then we should incorporate the ideas into the main text. E.g. instead of "Before we explore everything that MLEM can do, you need to save a machine learning model with MLEM." ([here](https://mlem.ai/doc/get-started?tab=Command-Line#saving-your-model)) we can use an explanation similar to the text above.

But if we put it in a hidden section that comes after another hidden section, to me that signals that we already decided this is not really important for the GS -- and I would agree it's best to keep the GS actionable (not theoretical). Thus moved it to a longer-format doc (that's linked several times from this page).

> putting an example

If I understand correctly this example would be disconnected from the main story in the GS which complicates things. We could def. add that example in the UG though, as another way to [expand on it](https://github.com/iterative/mlem.ai/pull/265#discussion_r1047651762).

Reviewer 2: I suggest to move it away from GS in my recent GS PR here: https://github.com/iterative/mlem.ai/pull/281
`user-guide/basic-concepts` seems to make sense to me

Author: So the change in this PR seems to point in the right direction. We have to merge one PR at a time so I'll focus on this one for now but I'm happy to solve any conflicts that arise in #281, unless yo guys think it's easier to do #281 first (lmk).

print(y_pred)
```

You can run it to print the prediction probabilities:

```cli
$ python predict.py
[[0.47 0.24 0.29]]
```

</tab>

<tab title="Command line">

### Batch scoring

`mlem` lets you use any saved model natively for prediction or batch scoring
with any dataset. This is very handy if you want to get quick feedback about a
model, for example. Let's create a sample dataset and apply it to the `rf` model
we saved earlier:

> ```csv
> sepal length (cm),sepal width (cm),petal length (cm),petal width (cm)
> 0,1,2,3

Reviewer: The fact it's csv is not reflected and it took me some time to understand what is this. I'd suggest (part 1/2)

```suggestion
```csv
# new_data.csv
sepal length (cm),sepal width (cm),petal length (cm),petal width (cm)
0,1,2,3
```

Author: Agree. I'll move the file name to the previous paragraph because `csv` doesn't have comments like that (looks odd)...

```

> Save the contents above in a file called `new_data.csv` and `mlem apply` itit to

Reviewer: ```suggestion
Save the contents above in a file and `mlem apply` it to
```

our model's `predict_proba` method as shown below to print the prediction
probabilities:

```cli
$ mlem apply models/rf new_data.csv \
    --method predict_proba \
    --import --import-type "pandas[csv]"
⏳️ Importing object from new_data.csv
⏳️ Loading model from models/rf.mlem
🍏 Applying `predict_proba` method...
[[0.47 0.24 0.29]]
```

<details>

> #### Click to learn more about the command options used.

Reviewer: `.` at the end looks strange to me - you suggest to click on the button, but the sentence ends with the dot. If this is our standard, you can ignore my opinion though

Author: Full sentences should end in `.` normally 🙂 

<img width="552" alt="image" src="https://user-images.githubusercontent.com/1477535/209634682-47516e41-dfc7-4b95-812e-4fa01b2e14a5.png">

> https://dvc.org/doc/start

<img width="1012" alt="image" src="https://user-images.githubusercontent.com/1477535/209634754-75779629-8bef-4694-835a-7e125478c048.png">

> https://dvc.org/doc/start/data-management/data-versioning

Ignoring 😬

> 🖇️  Adding route for /sklearn_predict
> 🖇️  Adding route for /sklearn_predict_proba

Reviewer: ```suggestion
🖇️  Adding route for /predict
🖇️  Adding route for /predict_proba
```
fix

Author: Thanks

Checkout openapi docs at <http://0.0.0.0:8080/docs>
...
INFO:     Uvicorn running on http://0.0.0.0:8080 (Press CTRL+C to quit)
```

> The server is now running and listening for requests on `localhost`. Endpoints

Reviewer: ```suggestion
The server is now running and listening for requests on `localhost:8080`. Endpoints
```

[Swagger UI](http://localhost:8080/docs) in your browser to explore the OpenAPI
spec and query examples.

[server implementations]: /doc/user-guide/serving

<admon type="tip">

MLEM can also [generate client-side] code to infer and query the model server

[generate client-side]: /doc/user-guide/serving#making-requests

</admon>

### Deploying models to production

MLEM lets you easily deliver your models for testing, shadowing, or production
with a variety of platforms like [Docker], [Heroku], [Sagemaker] and
[Kubernetes] (see also the [Deploying] guide). Now you don't have to deal with
low-level DevOps yourself. Let's look at a few examples:

[docker]: /doc/user-guide/deploying/Docker
[heroku]: /doc/user-guide/deploying/heroku
[sagemaker]: /doc/user-guide/deploying/sagemaker
[kubernetes]: /doc/user-guide/deploying/kubernetes
[deploying]: /doc/user-guide/deploying

<toggle>
<tab title="Docker container">

### Building and running a Dockerized ML model server

Let's first use the `mlem build` command to build a containerized model server.
With this single command, you build a FastAPI server packaged into a Docker
container image:

```cli

> $ mlem build Docker Docker-builder.mlem \

Reviewer: ```suggestion
$ mlem build docker docker-builder.mlem \
```

    --model models/rf \
    --image.name mlem-model
⏳️ Loading model from models/rf.mlem
...
✅  Built Docker image mlem-model:latest
```

This creates a `mlem-model:latest` Docker image and a builder metafile called

> `Docker-builder.mlem` describing the container.

Reviewer: ```suggestion
`docker-builder.mlem` describing the container.
```

Author: Oops. I capitalized these in bulk. Thanks for fix!

<details>

> #### Click to see to see the contents of `Docker-builder.mlem`.

Reviewer: ```suggestion
#### Click to see to see the contents of `docker-builder.mlem`.
```

```cli
$ mlem deployment run heroku app.mlem \
  --model models/rf \
  --app_name {myapp}-mlem-get-started-app
⏳️ Loading model from models/rf.mlem
...
  🛠 Building Docker image registry.heroku.com/{myapp}-mlem-get-started-app/web...
  ✅  Built Docker image registry.heroku.com/{myapp}-mlem-get-started-app/web
  🔼 Pushing image registry.heroku.com/{myapp}-mlem-get-started-app/web to registry.heroku.com
  ✅  Pushed image registry.heroku.com/{myapp}-mlem-get-started-app/web to registry.heroku.com
🛠 Releasing app {myapp}-mlem-get-started-app formation
✅  Service {myapp}-mlem-get-started-app is up. You can check it out at https://{myapp}-mlem-get-started-app.herokuapp.com/
```

A Deployment specification (or [declaration](/doc/command-reference/declare))
was written to the `app.mlem` metafile. You can use it to re-deploy the same
Heroku app with a different model.

<details>

#### Click to see the contents of `app.mlem`.

```yaml
app_name: example-mlem-get-started-app
object_type: deployment
type: heroku
```

<admon type="info">

Another special file was created: `app.mlem.state`. It contains metadata about
the deployment, including which MLEM model we used, the URL of the deployment,

> and other useful information ([more info]).

Reviewer: I'd delete that as well, but for now that file matters a lot and it's absence may cause problems. I assume we can remove it since the problem will be fixed https://github.com/iterative/mlem/issues/463
